File: artifacts/ht/translate/src/perfget.rs
use translate_new::*;
use std::env;
use std::process;
use std::fs::File;
use std::io::Read;
use std::io::Seek;
use std::io::SeekFrom;
use std::time::Instant;
use std::time::SystemTime;
use lazy_static::lazy_static;
use std::sync::Mutex;
lazy_static! {
    pub static ref FOUND: Mutex<Option<Box<dyn std::any::Any + Send>>> = Mutex::new(None);
}

pub fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("usage: perftest file");
        return Ok(());
    }

    let mut f = match File::open(&args[1]) {
        Ok(file) => file,
        Err(e) => {
            eprintln!("can't open file: {}: {}", args[1], e);
            return Ok(());
        }
    };

    let size = f.seek(SeekFrom::End(0))?;
    f.seek(SeekFrom::Start(0))?;

    let mut contents = vec![0; size as usize + 1];
    let nread = f.read(&mut contents[..size as usize])?;

    if nread != size as usize {
        eprintln!("read {} bytes instead of {}", nread, size);
        return Ok(());
    }

    contents[size as usize] = 0;

    let mut counts = match ht_create() {
        Some(ht) => ht,
        None => {
            exit_nomem();
            return Ok(());
        }
    };

    let mut p_idx = 0;
    let contents_str = unsafe { std::str::from_utf8_unchecked(&contents) };
    let mut p = contents_str;

    while p_idx < p.len() {
        while p_idx < p.len() && p.as_bytes()[p_idx] <= b' ' {
            p_idx += 1;
        }

        let word_start = p_idx;
        while p_idx < p.len() && p.as_bytes()[p_idx] > b' ' {
            p_idx += 1;
        }

        let word = &p[word_start..p_idx];
        if p_idx < p.len() {
            // In C, this would null-terminate, but in Rust we just slice
        }

        if let Some(entry) = ht_get(&counts, Some(word)) {
            if let Some(pcount) = entry.value.as_ref().and_then(|v| v.downcast_ref::<i32>()) {
                let pcount = *pcount + 1;
                let new_count = Box::new(pcount);
                if ht_set(&mut counts, Some(word), Some(new_count as Box<dyn std::any::Any>)).is_none() {
                    exit_nomem();
                    return Ok(());
                }
                continue;
            }
        }

        let pcount = Box::new(1);
        if ht_set(&mut counts, Some(word), Some(pcount as Box<dyn std::any::Any>)).is_none() {
            exit_nomem();
            return Ok(());
        }
    }

    let keys_len = ht_length(&counts);
    let mut keys = Vec::with_capacity(keys_len);
    let mut it = ht_iterator(&counts);
    let mut i = 0;
    while ht_next(&mut it) {
        if let Some(key) = &it.key {
            keys.push(key.clone());
            i += 1;
        }
    }

    let runs = 10;
    let start = Instant::now();
    for _run in 0..runs {
        for i in 0..keys_len {
            let key = &keys[i];
            if let Some(entry) = ht_get(&counts, Some(key)) {
                if let Some(entry_value) = entry.value.as_ref() {
                    if let Some(pcount) = entry_value.downcast_ref::<i32>() {
                        *FOUND.lock().unwrap() = Some(Box::new(*pcount) as Box<dyn std::any::Any + Send>);
                    }
                }
            }
        }
    }
    let end = Instant::now();
    let elapsed_ms = (end - start).as_secs_f64() * 1000.0;
    println!("{} runs getting {} keys: {:.03}ms", runs, keys_len, elapsed_ms);

    Ok(())
}

File: artifacts/ht/refactor/perfget.c
#include <ctype.h>
#include <ht.h>
#include <perfget.c>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct ht ht;
ht *ht_create(void);
void ht_destroy(ht *table);
void *ht_get(ht *table, const char *key);
const char *ht_set(ht *table, const char *key, void *value);
size_t ht_length(ht *table);
typedef struct 
{
  const char *key;
  void *value;
  ht *_table;
  size_t _index;
} hti;
hti ht_iterator(ht *table);
bool ht_next(hti *it);
void exit_nomem(void)
{
  fprintf(stderr, "out of memory\n");
  exit(1);
}

void *found;
int main(int argc, char **argv)
{
  if (argc < 2)
  {
    fprintf(stderr, "usage: perftest file\n");
    return 1;
  }
  FILE *f = fopen(argv[1], "rb");
  if (f == 0)
  {
    fprintf(stderr, "can't open file: %s\n", argv[1]);
    return 1;
  }
  fseek(f, 0, 2);
  long size = ftell(f);
  fseek(f, 0, 0);
  char *contents = (char *) malloc(size + 1);
  if (contents == 0)
  {
    exit_nomem();
  }
  size_t nread = fread(contents, 1, size, f);
  if (((long) nread) != size)
  {
    fprintf(stderr, "read %ld bytes instead of %ld", (long) nread, size);
    return 1;
  }
  fclose(f);
  contents[size] = 0;
  ht *counts = ht_create();
  if (counts == 0)
  {
    exit_nomem();
  }
  for (char *p = contents; p[p_idx]; p_idx += 1)
  {
    while (p[p_idx] && (p[p_idx] <= ' '))
    {
      p_idx += 1;
    }

    char *word = p;
    while (p[p_idx] && (p[p_idx] > ' '))
    {
      p_idx += 1;
    }

    if (p[p_idx] != 0)
    {
      p[p_idx] = 0;
    }
    void *value = ht_get(counts, word);
    if (value != 0)
    {
      int *pcount = (int *) value;
      *pcount += 1;
      continue;
    }
    int *pcount = malloc(sizeof(int));
    if (pcount == 0)
    {
      exit_nomem();
    }
    *pcount = 1;
    if (ht_set(counts, word, pcount) == 0)
    {
      exit_nomem();
    }
  }

  unsigned int p_idx = 0;
  const char **keys = malloc(ht_length(counts) * (sizeof(char *)));
  if (keys == 0)
  {
    exit_nomem();
  }
  hti it = ht_iterator(counts);
  int i = 0;
  while (ht_next(&it))
  {
    keys[i] = it.key;
    i += 1;
  }

  int runs = 10;
  clock_t start = clock();
  for (int run = 0; run < runs; run += 1)
  {
    for (int i = 0; i < ht_length(counts); i += 1)
    {
      found = ht_get(counts, keys[i]);
    }

  }

  clock_t end = clock();
  double elapsed_ms = (((double) (end - start)) / CLOCKS_PER_SEC) * 1000;
  printf("%d runs getting %lu keys: %.03fms\n", runs, ht_length(counts), elapsed_ms);
  return 0;
}


typedef struct ht ht;
ht *ht_create(void);
void ht_destroy(ht *table);
void *ht_get(ht *table, const char *key);
const char *ht_set(ht *table, const char *key, void *value);
size_t ht_length(ht *table);
typedef struct 
{
  const char *key;
  void *value;
  ht *_table;
  size_t _index;
} hti;
hti ht_iterator(ht *table);
bool ht_next(hti *it);
void exit_nomem(void)
{
  fprintf(stderr, "out of memory\n");
  exit(1);
}

void *found;
int main(int argc, char **argv)
{
  if (argc < 2)
  {
    fprintf(stderr, "usage: perftest file\n");
    return 1;
  }
  FILE *f = fopen(argv[1], "rb");
  if (f == 0)
  {
    fprintf(stderr, "can't open file: %s\n", argv[1]);
    return 1;
  }
  fseek(f, 0, 2);
  long size = ftell(f);
  fseek(f, 0, 0);
  char *contents = (char *) malloc(size + 1);
  if (contents == 0)
  {
    exit_nomem();
  }
  size_t nread = fread(contents, 1, size, f);
  if (((long) nread) != size)
  {
    fprintf(stderr, "read %ld bytes instead of %ld", (long) nread, size);
    return 1;
  }
  fclose(f);
  contents[size] = 0;
  ht *counts = ht_create();
  if (counts == 0)
  {
    exit_nomem();
  }
  for (char *p = contents; p[p_idx]; p_idx += 1)
  {
    while (p[p_idx] && (p[p_idx] <= ' '))
    {
      p_idx += 1;
    }

    char *word = p;
    while (p[p_idx] && (p[p_idx] > ' '))
    {
      p_idx += 1;
    }

    if (p[p_idx] != 0)
    {
      p[p_idx] = 0;
    }
    void *value = ht_get(counts, word);
    if (value != 0)
    {
      int *pcount = (int *) value;
      *pcount += 1;
      continue;
    }
    int *pcount = malloc(sizeof(int));
    if (pcount == 0)
    {
      exit_nomem();
    }
    *pcount = 1;
    if (ht_set(counts, word, pcount) == 0)
    {
      exit_nomem();
    }
  }

  unsigned int p_idx = 0;
  const char **keys = malloc(ht_length(counts) * (sizeof(char *)));
  if (keys == 0)
  {
    exit_nomem();
  }
  hti it = ht_iterator(counts);
  int i = 0;
  while (ht_next(&it))
  {
    keys[i] = it.key;
    i += 1;
  }

  int runs = 10;
  clock_t start = clock();
  for (int run = 0; run < runs; run += 1)
  {
    for (int i = 0; i < ht_length(counts); i += 1)
    {
      found = ht_get(counts, keys[i]);
    }

  }

  clock_t end = clock();
  double elapsed_ms = (((double) (end - start)) / CLOCKS_PER_SEC) * 1000;
  printf("%d runs getting %lu keys: %.03fms\n", runs, ht_length(counts), elapsed_ms);
  return 0;
}

